README.md

[![Arduino CI](https://github.com/RobTillaart/SET/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/SET/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/SET/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/SET/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/SET/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/SET/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/SET.svg?maxAge=3600)](https://github.com/RobTillaart/SET/releases)


# SET

Arduino library to implement a simple SET data structure.


## Description

The set library implements the set data structure for integers 0..255.
This limit is chosen because of the memory limitations of an Arduino UNO, 
however these numbers can be used as indices to a table of strings or other
data types.


## Interface


### Constructor

- **Set(bool clear = true)** creates an empty set, default it is cleared.
- **Set(&Set)** copies a set.


### Set level

- **clear()** empty the set.
- **invert()** flip all elements in the set.
- **addAll()** add all 256 elements to the set.
- **count()** returns the number of elements.
- **isEmpty()** idem
- **isFull()** idem


### Element level

- **add(uint8_t value)** add element n to the Set.
- **sub(uint8_t value)** remove element n from the Set.
- **invert(uint8_t value)** flip element n in the Set.
- **has(uint8_t value)** check if element n is in the Set.


### Operators

- union + +=
- diff - -=
- intersection * *=


### Equality

- equal ==
- not equal !=
- is subSet <=

A superSet B is not implemented as one could say B subSet A (B <= A)


### Iterators 

all iterator-functions returns the current element or -1 if not exist.

- **setCurrent(n)** if n is in the Set, n will be the current
- **first()** find the first element
- **next()** find the next element. Will not wrap around when 'end' of the set is reached.
- **prev()** find the previous element. Will not wrap around when 'begin' of the set is reached.
- **last()** find the last element.
- **getNth(n)** find the Nth element in a set if it exist.


## Operational

See examples


## Future

- update documentation
- 



CHANGELOG.md
# Change Log RunningMedian

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.2.6] - 2022-10-28
- Add RP2040 support to build-CI.
- Add CHANGELOG.md


## [0.2.5] - 2021-12-28
- update library.json
- update readme
- update license
- minor edits

## [0.2.4] - 2021-05-06
- getNth(n)

## [0.2.3] - 2021-05-05
- Add addAll (256 elements)
- setCurrent

## [0.2.2] - 2021-01-07
- Arduino-CI, unit test

## [0.2.1] - 2020-06-19
- fix library.json

## [0.2.0] - 2020-05-02
- refactored
- removed pre 1.0 support

----

## [0.1.11] - 2017-07-16
- fix count() --> 16 bit when set is full !

## [0.1.10] - 2017-07-16
- performance refactor
- isEmpty()

## [0.1.09] - 2015-07-12
- add const + constructor

## [0.1.08]
- memset for clr()

## [0.1.07]
- faster first/next/last/prev; interface

## [0.1.06]
- added flag to constructor to optimize +,-,*,
- set -> Set

## [0.1.05]
- bug fixing + performance a.o. count()

## [0.1.04]
- support for + - \*, some optimizations

## [0.1.03]
- changed &= to \*= to follow Pascal conventions

## [0.1.02]
- documentation

## [0.1.01]
- extending/refactor etc (09/11/2014)

## [0.1.00]
- initial version by Rob Tillaart (09/11/2014)
//

- initial version



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2021-01-07
// PURPOSE: unit tests for the SET library
//          https://github.com/RobTillaart/SET
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)

#include <ArduinoUnitTests.h>


#include "set.h"



unittest_setup()
{
}

unittest_teardown()
{
}


unittest(test_constructor)
{
  fprintf(stderr, "VERSION: %s\n", SET_LIB_VERSION);

  Set A;

  assertTrue(A.isEmpty());
  assertFalse(A.isFull());
  assertEqual(0, A.count());

  A.invert();
  assertFalse(A.isEmpty());
  assertTrue(A.isFull());
  assertEqual(256, A.count());

  A.clear();
  assertTrue(A.isEmpty());
  assertFalse(A.isFull());
  assertEqual(0, A.count());

  A.add(42);
  assertFalse(A.isEmpty());
  assertFalse(A.isFull());
  assertEqual(1, A.count());

  A.addAll();
  assertFalse(A.isEmpty());
  assertTrue(A.isFull());
  assertEqual(256, A.count());
}


unittest(test_count_has)
{
  fprintf(stderr, "VERSION: %s\n", SET_LIB_VERSION);

  Set A;

  A.clear();
  for (int i = 0; i < 10; i++)
  {
    assertEqual(i, A.count());
    A.add(i);
  }

  for (int i = 0; i < 10; i++)
  {
    assertTrue(A.has(i));
  }

  for (int i = 0; i < 10; i++)
  {
    assertEqual(10 - i, A.count());
    A.sub(i);
  }
  assertEqual(0, A.count());

}


unittest(test_operator)
{
  fprintf(stderr, "VERSION: %s\n", SET_LIB_VERSION);

  Set A;
  Set B;
  Set C;

  A.clear();
  B.clear();
  for (int i = 0; i < 10; i++)
  {
    A.add(i);
    B.add(i + 7);
  }
  C = A + B;
  assertEqual(10, A.count());
  assertEqual(10, B.count());
  assertEqual(17, C.count());

  C = A - B;
  assertEqual(7, C.count());

  C = A * B;
  assertEqual(3, C.count());

  C = A;
  A += B;
  assertEqual(17, A.count());

  A -= B;
  assertEqual(7, A.count());

  A *= B;
  assertEqual(0, A.count());

  C *= B;
  assertTrue( A == A );
  assertTrue( A != B );
  assertTrue( C <= B);
}


unittest(test_iterator)
{
  fprintf(stderr, "VERSION: %s\n", SET_LIB_VERSION);

  Set A;

  A.clear();

  for (int i = 0; i < 20; i++)
  {
    A.add(i*3);
  }
  assertEqual(20, A.count());

  int cur = A.first();
  for (int i = 0; i < 20; i++)
  {
    assertEqual(i * 3, cur);
    cur = A.next();
  }
  
  cur = A.last();
  for (int i = 20; i > 0; i--)
  {
    assertEqual(i * 3, cur);
    cur = A.prev();
  }
  
  assertEqual(0, A.getNth(1));
  assertEqual(3, A.getNth(2));
  assertEqual(6, A.getNth(3));
  assertEqual(9, A.getNth(4));
  assertEqual(12, A.getNth(5));
  assertEqual(15, A.getNth(6));
  assertEqual(18, A.getNth(7));
  assertEqual(21, A.getNth(8));

}

unittest_main()

// --------


set.h
#pragma once
//
//    FILE: set.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.2.4
//    DATE: 2014-09-11
// PURPOSE: SET library for Arduino
//     URL: https://github.com/RobTillaart/SET


#include "Arduino.h"


#define SET_LIB_VERSION         (F("0.2.4"))


class Set
{
public:
    explicit Set(const bool clear = true);     // create empty Set
    Set(const Set &t);                // create copy Set

    void     clear();                 // clear the Set
    void     clr() { clear(); };      // will become obsolete 0.3.0
    void     invert();                // flip all elements in the Set
    void     addAll();                // add all elements
    uint16_t count() const;           // return the #elements
    bool     isEmpty();
    bool     isFull();


    void add(const uint8_t value);     // add element to the Set
    void sub(const uint8_t value);     // remove element from Set
    void invert(const uint8_t value);  // flip element in Set
    bool has(const uint8_t value);     // element is in Set


    Set operator + (const Set &);     // union
    Set operator - (const Set &);     // diff
    Set operator * (const Set &);     // intersection

    void operator += (const Set &);   // union
    void operator -= (const Set &);   // diff
    void operator *= (const Set &);   // intersection


    bool operator == (const Set &) const;   // equal
    bool operator != (const Set &) const;   // not equal
    bool operator <= (const Set &) const;   // is subSet,
    // a superSet b is not implemented as one could 
    // say b subSet a (b <= a)
    // a <= b  


    // iterating through the Set
    // returns value or -1 if not exist
    int setCurrent(const uint8_t current);  // set element as current
    int first();                            // find first element
    int next();                             // find next element
    int prev();                             // find previous element
    int last();                             // find last element
    int getNth(const uint8_t n);            // find Nth element in a set (from start)

private:
    uint8_t _mem[32];           // can hold 0..255
    uint8_t _masks[8] = {1, 2, 4, 8, 16, 32, 64, 128};
    int _current = -1;
	
    int findNext(const uint8_t p, const uint8_t q); // helper for first, next
    int findPrev(const uint8_t p, const uint8_t q); // helper for last, prev
};


// -- END OF FILE --



set.cpp
//
//    FILE: set.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.2.4
//    DATE: 2014-09-11
// PURPOSE: SET library for Arduino
//     URL: https://github.com/RobTillaart/SET
//
//  HISTORY:
//  0.2.4   2021-05-06  getNth(n) 
//  0.2.3   2021-05-05  Add addAll (256 elements) + setCurrent
//  0.2.2   2021-01-07  Arduino-CI, unit test
//  0.2.1   2020-06-19  fix library.json
//  0.2.0   2020-05-02  refactored, removed pre 1.0 support
//  0.1.11  2017-07-16  fix count() --> 16 bit when set is full !
//  0.1.10  2017-07-16  performance refactor. isEmpty()
//  0.1.09  2015-07-12  const + constructor
//  0.1.08              memset for clr()
//  0.1.07  faster first/next/last/prev; interface
//  0.1.06  added flag to constructor to optimize +,-,*,
//          set -> Set
//  0.1.05  bug fixing + performance a.o. count()
//  0.1.04  support for + - *, some optimizations
//  0.1.03  changed &= to *= to follow Pascal conventions
//  0.1.02  documentation
//  0.1.01  extending/refactor etc (09/11/2014)
//  0.1.00  initial version by Rob Tillaart (09/11/2014)
//


#include "set.h"

/////////////////////////////////////////////////////
//
// CONSTRUCTORS
//
Set::Set(const bool clear)
{
    if (clear)
    {
        clr();
    }
    _current = -1;
}


Set::Set(const Set &t)
{
    for (uint8_t i = 0; i < 32; i++)
    {
        _mem[i] = t._mem[i];
    }
    _current = -1;
}


/////////////////////////////////////////////////////
//
// METHODS
//
void Set::add(const uint8_t value)
{
    uint8_t idx = value / 8;
    _mem[idx] |= _masks[value & 7];
}


void Set::sub(const uint8_t value)
{
    uint8_t idx = value / 8;
    _mem[idx] &= ~_masks[value & 7];
}


void Set::invert(const uint8_t value)
{
    uint8_t idx = value / 8;
    _mem[idx] ^= _masks[value & 7];
}


void Set::addAll()
{
    memset(_mem, 0xFF, 32);
}


bool Set::has(const uint8_t value)
{
    uint8_t idx = value / 8;
    return (_mem[idx] & _masks[value & 7]) > 0;
}


uint16_t Set::count() const
{
    uint16_t cnt = 0;
    
    uint8_t i = 32;
    do
    {
        // kerningham bit count trick
        uint8_t b = _mem[--i];
        for (; b; cnt++)
        {
            b &= b-1;
        }
    }
    while (i != 0);
    return cnt;
}


void Set::clear()
{
    memset(_mem, 0, 32);
}


void Set::invert()
{
    uint8_t i = 32;
    do
    {
        _mem[--i] ^= 0xFF;
    } 
    while (i != 0);
}


bool Set::isEmpty()
{
    uint8_t i = 32;
    do
    {
      if (_mem[--i] > 0) return false;
    } 
    while (i != 0);
    return true;
}


bool Set::isFull()
{
    // check two elements per loop 
    // is faster for full sets but slower for empty set.
    // footprint is ~25 bytese larger
    // overal performance gain 
    uint8_t i = 32;
    do
    {
        if ((_mem[--i]) != 255) return false;
    } 
    while (i != 0);
    return true;
}


int Set::setCurrent(const uint8_t current)
{
    _current = -1;
    if (has(current))
    {
      _current = current;
    }
    return _current;
}


int Set::first()
{
    if (has(0))
    {
      _current = 0;
      return _current;
    }
    return findNext(0, 0);
}


int Set::next()
{
    if (_current & 0x8000) return -1; // if current == -1
    _current++;
    uint8_t p = (uint8_t)_current / 8;
    uint8_t q = (uint8_t)_current & 7;
    return findNext(p, q);
}

// pointer math version ~12% faster but not for previous
// needs investigation.
// int Set::findNext(const uint8_t p, const uint8_t q)
// {
	// uint8_t * pp = &_mem[p];
	// uint8_t mask = 1 << q;
	// uint8_t j = q;
	// do
	// {
		// if (*pp != 0)
		// {
            // while (j < 8)
            // {
                // if (*pp & mask)
                // {
                    // _current = (pp - _mem) * 8 + j;
                    // return _current;
                // }
                // mask <<= 1;
				// j++;
            // }			
		// }
		// j = 0;
		// mask = 1;
		// pp++;
	// } 
	// while (pp != &_mem[31]);
    // _current = -1;
    // return _current;
// }


int Set::findNext(const uint8_t p, uint8_t q)
{
    for (uint8_t i = p; i < 32; i++)
    {
        uint8_t b = _mem[i];
        if (b != 0)
        {
            uint8_t mask = 1 << q;  // _masks[q]
            for (uint8_t j = q; j < 8; j++)
            {
                if (b & mask)
                {
                    _current = i * 8 + j;
                    return _current;
                }
                mask <<= 1;
            }
        }
        q = 0;
    }
    _current = -1;
    return _current;
}


int Set::prev()
{
    if (_current & 0x8000) return -1;
    _current--;
    uint8_t p = (uint8_t)_current / 8;
    uint8_t q = (uint8_t)_current & 7;
    return findPrev(p, q);
}


int Set::last()
{
    if (has(255))
    {
      _current = 255;
      return _current;
    }
    return findPrev(31, 7);
}


int Set::getNth(const uint8_t n)
{
    if (n == 0) return -1;
    if (n == 1) return first();
    _current = first();
    int i = 1;
    while ((_current > -1) && (i < n))
    {
      _current = next();
      i++;
    }
    return _current;
}


int Set::findPrev(const uint8_t p, uint8_t q)
{
    uint8_t m = 1 << q;
    for (uint8_t i = p; i != 255; --i)  // uint < 0
    {
        uint8_t b = _mem[i];
        if (b != 0)
        {
            uint8_t mask = m;
            for (uint8_t j = q; j != 255; --j)
            {
                if (b & mask)
                {
                    _current = i * 8 + j;
                    return _current;
                }
                mask >>= 1;
            }
        }
        m = 128;  // 1 << 7;
        q = 7;
    }
    _current = -1;
    return _current;
}


/////////////////////////////////////////////////////
//
// OPERATORS
//

Set Set::operator + (const Set &t)  // union
{
    Set s(false);
    for (uint8_t  i = 0; i < 32; i++)
    {
        s._mem[i] = this->_mem[i] | t._mem[i];
    }
    return s;
}


Set Set::operator - (const Set &t)  // diff
{
    Set s(false);
    for (uint8_t i = 0; i < 32; i++)
    {
        s._mem[i] = this->_mem[i] & ~t._mem[i];
    }
    return s;
}


Set Set::operator * (const Set &t)  // intersection
{
    Set s(false);
    for (uint8_t i = 0; i < 32; i++)
    {
        s._mem[i] = this->_mem[i] & t._mem[i];
    }
    return s;
}


void Set::operator += (const Set &t)  // union
{
    for (uint8_t i = 0; i < 32; i++)
    {
        _mem[i] |= t._mem[i];
    }
}


void Set::operator -= (const Set &t)  // diff
{
    for (uint8_t i = 0; i < 32; i++)
    {
        _mem[i] &= ~t._mem[i];
    }
}


void Set::operator *= (const Set &t)  // intersection
{
    for (uint8_t i = 0; i < 32; i++)
    {
        _mem[i] &= t._mem[i];
    }
}


bool Set::operator == (const Set &t) const // equal
{
    for (uint8_t i = 0; i < 32; i++)
    {
        if (_mem[i] != t._mem[i]) return false;
    }
    return true;
}


bool Set::operator != (const Set &t) const // not equal
{
    for (uint8_t i = 0; i < 32; i++)
    {
        if (_mem[i] != t._mem[i]) return true;
    }
    return false;
}


bool Set::operator <= (const Set &t) const // subSet
{
    for (uint8_t i = 0; i < 32; i++)
    {
        if ((_mem[i] & ~t._mem[i]) > 0) return false;
    }
    return true;
}


// -- END OF FILE --

